rest_framework_jwt/serializers.py

```diff
@@ -193,3 +193,29 @@ class DeviceSerializer(serializers.ModelSerializer):
     class Meta:
         model = Device
         fields = ['id', 'created', 'name', 'details', 'last_request_datetime']
+
+
+class DeviceTokenRefreshSerializer(Serializer):
+    permanent_token = serializers.CharField(required=True)
+
+    def validate(self, attrs):
+        permanent_token = attrs['permanent_token']
+        try:
+            device = Device.objects.get(permanent_token=permanent_token)
+        except Device.DoesNotExist:
+            raise serializers.ValidationError({'permanent_token': _('Invalid permanent_token value.')})
+
+        now = datetime.now()
+        if now > device.last_request_datetime + api_settings.JWT_PERMANENT_TOKEN_EXPIRATION_DELTA:
+            device.delete()
+            raise serializers.ValidationError({'permanent_token': _('Permanent token has expired.')})
+
+        if now > device.last_request_datetime + api_settings.JWT_PERMANENT_TOKEN_EXPIRATION_ACCURACY:
+            device.last_request_datetime = now
+            device.save()
+
+        payload = jwt_payload_handler(device.user, device=device)
+        return {
+            'token': jwt_encode_handler(payload),
+            'user': device.user
```

Comment:
Even if not, it should not return user, only token. And, btw, why you return manually data instead of using serializers?